pkg/gofr/datasource/mongo/mongo.go

```diff
@@ -77,27 +78,103 @@ func (c *Client) UseTracer(tracer any) {
 }
 
 // Connect establishes a connection to MongoDB and registers metrics using the provided configuration when the client was Created.
-func (c *Client) Connect() {
+func (c *Client) Connect(ctx context.Context) error {
 	c.logger.Logf("connecting to mongoDB at %v to database %v", c.config.URI, c.config.Database)
 
-	uri := c.config.URI
+	uri := c.getURI()
 
-	if uri == "" {
-		uri = fmt.Sprintf("mongodb://%s:%s@%s:%d/%s?authSource=admin",
-			c.config.User, c.config.Password, c.config.Host, c.config.Port, c.config.Database)
+	client, err := c.createClient(ctx, uri)
+	if err != nil {
+		return err
 	}
 
-	m, err := mongo.Connect(context.Background(), options.Client().ApplyURI(uri))
+	if err := c.pingDatabase(ctx, client); err != nil {
+		return err
+	}
+
+	c.setupMetrics()
+
+	c.Database = client.Database(c.config.Database)
+
+	return c.verifyDatabaseAccess(ctx)
+}
+
+func (c *Client) getURI() string {
+	if c.config.URI != "" {
+		return c.config.URI
+	}
+
+	return fmt.Sprintf("mongodb://%s:%s@%s:%d/%s?authSource=admin",
+		c.config.User, c.config.Password, c.config.Host, c.config.Port, c.config.Database)
+}
+
+func (c *Client) createClient(ctx context.Context, uri string) (*mongo.Client, error) {
+	clientOpts := options.Client().ApplyURI(uri)
+
+	client, err := mongo.Connect(ctx, clientOpts)
 	if err != nil {
-		c.logger.Errorf("error connecting to mongoDB, err:%v", err)
+		return nil, c.handleConnectionError(err)
+	}
+
+	return client, nil
+}
+
+func (c *Client) handleConnectionError(err error) error {
+	if c.isAuthenticationError(err) {
+		return fmt.Errorf("%w: %w", ErrAuthentication, err)
+	}
+
+	if c.isTimeoutError(err) {
+		return fmt.Errorf("%w: connection timeout", ErrGenericConnection)
+	}
+
+	return fmt.Errorf("%w: %w", ErrGenericConnection, err)
+}
+
+func (*Client) isTimeoutError(err error) bool {
+	return strings.Contains(err.Error(), "connection timeout") || mongo.IsTimeout(err)
+}
 
-		return
+func (*Client) isAuthenticationError(err error) bool {
+	return strings.Contains(err.Error(), "authentication failed") ||
+		strings.Contains(err.Error(), "AuthenticationFailed")
+}
+
+func (c *Client) pingDatabase(ctx context.Context, client *mongo.Client) error {
+	if err := client.Ping(ctx, nil); err != nil {
+		return c.handlePingError(err)
+	}
+
+	return nil
+}
+
+func (c *Client) handlePingError(err error) error {
+	if mongo.IsTimeout(err) {
+		return fmt.Errorf("%w: connection timeout", ErrGenericConnection)
```

Comment:
```suggestion
		return fmt.Errorf("%w: connection timeout: %w", ErrGenericConnection, err)
```

You should never hide low level error with yours. You should wrap them with yours